src/plugins/ucc_discovery_plugin.cpp

```diff
@@ -0,0 +1,345 @@
+
+  if (!root_node) {
+    // input node may already be nullptr in case we try to get right input of node with only one input
+    return nullptr;
+  } else {
+    visit_lqp(root_node, [&](auto& node) {
+      if (node->type != LQPNodeType::Predicate) {
+        return LQPVisitation::VisitInputs;
+      }
+
+      // when looking at predicate node, check whether the searched column is filtered in this predicate
+      // -> if so, it is a valid UCC candidate; if not, still continue search
+      const auto casted_node = std::static_pointer_cast<PredicateNode>(node);
+
+      // first, ensure that we look at a binary predicate expression checking for equality (e.g., A==B)
+      const auto predicate = std::dynamic_pointer_cast<BinaryPredicateExpression>(casted_node->predicate());
+      if ((!predicate) || (predicate->predicate_condition != PredicateCondition::Equals)) {
+        return LQPVisitation::VisitInputs;
+      }
+
+      // get the column expression, should be left, but also check the right operand if the left one is not column
+      auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(predicate->left_operand());
+      auto value_expression = std::dynamic_pointer_cast<ValueExpression>(predicate->right_operand());
+      if (!column_expression) {
+        column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(predicate->right_operand());
+        value_expression = std::dynamic_pointer_cast<ValueExpression>(predicate->left_operand());
+      }
+
+      if ((!column_expression) || (!value_expression)) {
+        // predicate needs to look like column = value or value = column; if not, move on
+        return LQPVisitation::VisitInputs;
+      }
+
+      if (column_expression == column_candidate) {
+        // equal condition and join column -> candidate for UCC
+        const auto table = std::static_pointer_cast<const StoredTableNode>(column_expression->original_node.lock());
+        candidates->push_back(UCCCandidate{table->table_name, column_expression->original_column_id});
+        return LQPVisitation::VisitInputs;
+      }
+
+      const auto expression_table =
+          std::static_pointer_cast<const StoredTableNode>(column_expression->original_node.lock());
+      const auto candidate_table =
+          std::static_pointer_cast<const StoredTableNode>(column_candidate->original_node.lock());
+
+      if (expression_table == candidate_table) {
+        // both columns same table -> if both are UCC we could still convert join -> both are UCC candidate
+        candidates->push_back(UCCCandidate{expression_table->table_name, column_expression->original_column_id});
+        candidates->push_back(UCCCandidate{candidate_table->table_name, column_candidate->original_column_id});
+      }
+
+      return LQPVisitation::VisitInputs;
+    });
+  }
+
+  return candidates;
+}
+
+UCCCandidates UccDiscoveryPlugin::identify_ucc_candidates() const {
```

Comment:
This function is well over 100 lines long and also deeply nested. Please simplify.